### ISSUE TITLE
Dropdown: don't ignore keydown from input or textarea

### ISSUE THREAD

**Issue by gpakosz** [twbs/bootstrap#21366]
I'm reopening an issue for v4, as #15757 got [closed without resolution for v3](https://github.com/twbs/bootstrap/issues/15757#event-778679826).

At the time I opened the issue, [v4 was still filtering key down events for input or textarea](https://github.com/twbs/bootstrap/blob/3c12d73203bad27edcaa1072e263fdab2a2c14f7/js/src/dropdown.js#L214-L218):
```
static _dataApiKeydownHandler(event) {
  if (!/(38|40|27|32)/.test(event.which) ||
     /input|textarea/i.test(event.target.tagName)) {
    return
  }
```

In my opinion, a common use of an `<input type="text">` and dropdown combination would be displaying search results as a dropdown ala http://developer.github.com:
![image](https://cloud.githubusercontent.com/assets/553208/21317711/9c30916a-c606-11e6-8166-9e50e16ac71f.png)

In that respect, can you please make the behavior configurable (if you don't want to revert the default behavior).

Thank you.

**Comment by gpakosz** [comment]
About the original issue discussion, I disagree with @fat's assertion:
> ... and most people aren't putting inputs in their dropdowns to begin with) ...

A search box with dropdown results is not a rare pattern. This is what the search box does on https://twitter.com btw.

@hnrch02 I had a look at the [original comment](https://github.com/twbs/bootstrap/issues/15757#issuecomment-85457384) but in the corresponding JSBin, clicking the input element in the `.dropdown-menu` closes the dropdown right away and I don't get a chance to hit space. So I'm not sure what it was trying to demonstrate with spaces.

As far as key down/up are concerned: imho when `<input type="text">` lives inside `.dropdown-menu`, default behavior for key down/up should be to select next/previous item. So yes I believe this would solve it.

**Comment by hnrch02** [comment]
> A search box with dropdown results is not a rare pattern. This is what the search box does on https://twitter.com btw.

This pattern is not rare, I agree. However, this concerns input fields/textareas **inside** the dropdown menu itself, which I think is rather uncommon usage.

> @hnrch02 I had a look at the original comment but in the corresponding JSBin, clicking the input element in the .dropdown-menu closes the dropdown right away and I don't get a chance to hit space. So I'm not sure what it was trying to demonstrate with spaces.

You can compare [this Bin with v3.3.0](https://output.jsbin.com/yozopa/2) (where you can't type out spaces in the input) with [this one using v3.3.7](https://output.jsbin.com/yozopa/3).

> As far as key down/up are concerned: imho when <input type="text"> lives inside .dropdown-menu, default behavior for key down/up should be to select next/previous item. So yes I believe this would solve it.

Alright then.

**Comment by gpakosz** [comment]
> This pattern is not rare, I agree. However, **this concerns input fields/textareas inside the dropdown menu itself, which I think is rather uncommon usage**.

I get it and I'm sorry for the misunderstanding. I agree, an input/textarea **inside** the dropdown menu is rather uncommon.

All in all, doesn't this discussion mean that the handler should ultimately distinguish between:

  - key events on an input to which dropdown results are "attached" (almost like in 3.3.0):
    - space should add a space to the input
    - down/up should navigate the dropdown menu
      - in that respect the 3.3.0 bin linked above fails to navigate up from the first `<li>` to the `<input>` above

vs

  - key events on an input living inside a dropdown menu (like in 3.3.7):
    - space should add a space to the input
    - down/up should jump to beginning/end of input text


What do you think?

**Comment by vanduynslagerp** [comment]
Fix for the keyboard navigation: #21535 
Fix for handling input and textarea better: #21802
